content/workshops/micropython-in-browser/index.md

```diff
@@ -0,0 +1,614 @@
+client.connect()
+print("Connected to MQTT broker")
+```
+
+### Task 3: Initialize imu
+
+```python
+i2c = I2C(0, scl=Pin(8), sda=Pin(7), freq=400000)
+
+# Create IMU instance
+imu = ICM42670(i2c)
+
+# Configure sensor
+imu.configure(
+    gyro_fs=GyroFS.FS_500DPS,
+    gyro_odr=ODR.ODR_100_HZ,
+    accel_fs=AccelFS.FS_4G,
+    accel_odr=ODR.ODR_100_HZ
+)
+
+# Enable sensors
+imu.set_accel_power_mode(PowerMode.LOW_NOISE)
+imu.set_gyro_power_mode(PowerMode.LOW_NOISE)
+```
+
+### Task 4: Read Orientation and publish
+
+```python
+def publish_orientation():
+    """Read orientation and publish to MQTT"""
+
+    while True:
+        # Read sensor data
+        data = imu.get_all_data()
+        accel = (data['accel']['x'], data['accel']['y'], data['accel']['z'])
+        gyro = (data['gyro']['x'], data['gyro']['y'], data['gyro']['z'])
+
+        # Calculate orientation using complementary filter
+        angles = imu.complementary_filter(accel, gyro)
+        roll = angles['roll']
+        pitch = angles['pitch']
+
+        # Interpret orientation
+        if abs(pitch) < 20 and abs(roll) < 20:
+            position = "Flat"
+        elif pitch > 30:
+            position = "Tilted Forward"
+        elif pitch < -30:
+            position = "Tilted Backward"
+        elif roll > 30:
+            position = "Tilted Right"
+        elif roll < -30:
+            position = "Tilted Left"
+        else:
+            position = "Diagonal"
+
+        # Create message
+        message = f"Roll: {roll:.1f}°, Pitch: {pitch:.1f}° -> {position}"
+        print(message)
+
+        # Publish to MQTT
+        client.publish(TOPIC, position.encode())
+
+        time.sleep(0.5)
+
+publish_orientation()
+```
+
+## Conclusion
+
+You've learned to program ESP32 microcontrollers using browser-based Jupyter notebooks, covering:
+
+- MicroPython firmware flashing and interactive development
+- GPIO control for LEDs and button input
+- ESP-NOW wireless communication protocol
+- IMU sensor data acquisition and processing
+- MQTT protocol for IoT messaging
+
+These skills form the foundation for building sophisticated IoT applications. The browser-based approach eliminates toolchain complexity while maintaining full access to MicroPython's capabilities.
+
+## FAQ
+- What kernel should I use?
+  - Select `Embedded Kernel`.
+- How do I connect my device to the Jupyter notebook?
+  - Click on the ESP Control Panel, click `Connect` and select your device.
+- Which of the devices is my ESP board?
+  - The ESP board usually appears as `USB JTAG`.
+- I cannot connect to a different Jupyter notebook.
+  - Disconnect from the previous notebook and connect the new one.
+- The notebook isn’t running my code, what should I do
+  - In the ESP Control Panel, click `Disconnect device`, then reopen the notebook and connect again.
+
+## Feedback
+If you have feedback for this workshop, please feel free to contact me at [daniel.paul@espressif.com](mailto:daniel.paul@espressif.com) 🚀
```

Comment:
Please do not share your email here. You can do it during the workshop, only for the participants.

Reply:
Removed, thank you for suggestion!